### ISSUE TITLE
Accessibility: Proper semantic HTML or `role` attributes

### ISSUE THREAD

**Issue by jhildenbiddle** [docsifyjs/docsify#2295]
## Feature request

Docsify does not use semantic HTML or assign `role` attributes properly. This introduces accessibility issues by preventing screen readers from properly identifying and navigating directly to important areas like the main content and site navigation.

Fortunately, these issues can be address easily and in a non-breaking way by assigning appropriate [WAI-ARIA roles](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles) to existing elements. If breaking changes are permitted, it would be preferable to revisit Docsify's DOM structure and use of HTML tags which would reduce the need to assign `role` attributes.

#### Problem or desire

- The `<main>` element is improperly used as a wrapper for the sidebar, the sidebar toggle button, and the main content. As a result, screen readers are are unable to navigate directly to the main page content.
- The sidebar navigation acts as the site's primary navigation, however it is not rendered with a `<nav>` element or assigned a `role` of `navigation`. As a result, screen readers are unable to navigate directly to the site's primary navigation.
- The top navigation is rendered using a `<nav>` element and no other element has been assigned a `role` of `navigation`. As a result, screen readers either 1) improperly identify the top navigation as the site's primary navigation or 2) are unable to identify and navigate directly to any site navigation when the top nav is not rendered.
- The site name/logo is contained with an `<h1>` element. As a result, screen readers incorrectly identify this is a level one heading instead of a 
- The progress bar is rendered as a `<div class="progress">` element. As a result, screen readers are unable to detect or announce progress updates.
- The cover page container lacks 

#### Proposal

For the main content:

- **Option 1** (breaking): Move sidebar and sidebar toggle elements outside of `<main>` element. Consider removing unnecessary wrapper elements like `<section class="content">` and using a different ID for `<article id="main">` or omitting the ID entirely to avoid confusion.
- **Option 2** (non-breaking): Add [`role="main"`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/main_role) to Docsify's `<section class="content">` or `<article id="main">` element.

For the sidebar / primary navigation: 

- **Option 1** (breaking): Replace the `<div class="sidebar-nav">` element with a [`<nav>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/nav) element.
- **Option 2** (non-breaking): Add [`role="navigation"`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/navigation_role) and to the `<div class="sidebar-nav">` element.

For the top navigation:

- **Option 1** (non-breaking): Add [`role="menubar"`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/menubar_role) and related aria attributes to the `<nav class="app-nav">` element. Optionally consider replacing the `<nav>` element with a generic `<div>`.

For the site name/logo:

- **Option 1** (breaking): Wrap the site name/log in a [`<header>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/header) instead of `<h1>`, then wrap the site name/logo and the search feature within a [`<banner>`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/banner_role) element.

For the progress bar:

- **Option 1** (breaking): Replace `<div class="progress">` with a [`<progress>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/progress) element.
- **Option 2** (non-breaking): Add [`role="progressbar"`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/progressbar_role) and related aria attributes to the `<div class="progress"> element.

For the cover page:

- **Option 1** (non-breaking): Add [`role="region"`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/region_role) and `aria-label="Cover page"` (or similar) attributes to the `<section class="cover">` element.

#### Implementation

See above.

**Comment by trusktr** [comment]
These will be really great improvements! The breaking changes would be more ideal in terms of code quality.

Maybe we can release this in such a way so that people not using versioned URLs will not get the update, by f.e. releasing the next major version on a new path? For example:

- someone currently using `//cdn.jsdelivr.net/npm/docsify` will get redirected to the latest main file, which will be v4.
- with the update, we add a console message into v4 that tells people if they want to update to v5 (or vX) they need to opt-in with a specific path, f.e. `//cdn.jsdelivr.net/npm/docsify/dist/docsify.5.js`
- As we release new versions, each major release has a new path and keeps the last N versions, that way we can force people to use versions if they want to stay behind, and people who stopped maintaining their sites and didn't use versioned URLs never have a broken site.

The downside of this is that we will be publishing old and new versions. F.e. 5.0 will have 4 and 5. 6.0 will have 4, 5, and 6. Etc.

What if we just keep the last two or three versions, and delete a version after it has been logging a message to console for a year (or some time value)?

F.e. a user who currently does not use versioned URLs will start to see something like this in console:

> Important! If you see this message, you are on Docsify v4 and you are not using a version number in the Docsify script URL. This script will stop being available at this location on or after Nov 22, 2024. To prevent your site from breaking, add a version to your URL like so: ....... Or better yet, update to the latest Docsify path....

**Comment by jhildenbiddle** [comment]
> These will be really great improvements!

Thanks! They're implemented and ready for review in #2304. 😄 

> The breaking changes would be more ideal in terms of code quality.

I opted to implement the above enhancements as non-breaking changes for the following reasons:

1. The "breaking" options above will impact third party plugins and themes, which we want to avoid for v5 if the intention is for v5 to contain only "legacy-breaking" changes.
2. Updating the UI (modernize, redesign, componentize, or all of the above) seems like a logical enhancement to target next. Those UI changes will obviously introduce breaking changes for plugins and themes. I think the community will be better served by incorporating the "breaking" accessibility enhancement above into the "breaking" future UI updates so we can minimize ecosystem-breaking changes. This will allow us to deliver all planned "breaking" UI changes in a single release.

I realize discussions about what comes *after* v5 may introduce eye rolls given how long it's taken us to get anywhere near a v5 release, but I believe we can keep the next two releases narrow in scope and release them in a reasonable time frame:

**v5**

- Accessibility enhancements (non-breaking)
- Drop legacy support (breaking)
- New v5 CDN files available in `/dist`. Docs recommend `@5` version lock in URLs.
- New v4 redirect files available in `/lib` redirecting to `@4/lib` CDN files.
- Old v4 CDN files available via `@4/lib` (forever).
- Old v4 CDN files available via `@latest/lib` (temporary). 
- Console warning displayed when `@latest` Docsify CDN URL detected.

**v6**

- UI update: modernization, redesign, possibly componentize (breaking)
- Plugin system update (breaking) (see #1209 for previous discussion)
- New v6 CDN files available in `/dist`. Docs recommend `@6` version lock in URLs.
- Old v5 CDN files available via `@5/dist` (forever).
- Old v4 CDN removed.

**Comment by trusktr** [comment]
Sounds good. I think we can go with the non-ecosystem-breaking approach first, and keep good notes on how to convert that to the better format in the ecosystem-breaking change later.

I know that CDNs support versions numbers in the URLs, but for websites that have no version numbers, they'll break if we eventually remove the v4 paths. Perhaps v4 is the only file path we support for a while, and new paths have can come and go assuming that our docs clearly tell people the new path to use.

I need to refresh on the output format and come back with a concrete plan of action for it (or do you already have one?)

**Comment by jhildenbiddle** [comment]
> ...for websites that have no version numbers, they'll break if we eventually remove the v4 paths.

Right. The proposal above will break v4 sites without an `@` version lock when v6 is released. That's just a suggestion. We can look at [CDN statistics](https://www.jsdelivr.com/package/npm/docsify?tab=stats) when the time comes to help determine if/when we remove 4v paths and redirects.

It's worth mentioning that publishing *any* breaking change will likely break old v4 sites since there was a time when our [documentation](https://github.com/docsifyjs/docsify/blob/7c6ab54ba7b9701daa73eb929dce7c8e7b049eca/docs/quickstart.md) recommended loading the latest docsify version without a version lock or a path:

```html
<script src="//unpkg.com/docsify"></script>
```

This was a long time ago though and I don't know of any workarounds worth considering.

> I need to refresh on the output format and come back with a concrete plan of action for it (or do you already have one?)

Haven't thought about it much, but I assume we would want to offer both an IIFE (as we do today) and an ES module.

Let's continue to conversation in Discord or a [public discussion](https://github.com/docsifyjs/docsify/discussions) since this issue is closed.
